### PR TITLE
Adding ldap_entry module

### PR DESCRIPTION
**Pull request by jtyr**
##### ISSUE TYPE

New Module Pull Request
##### COMPONENT NAME

`ldap_entry`
##### ANSIBLE VERSION

For Ansible 2.3
##### SUMMARY

This PR is adding a new module that allows to add/change/remove LDAP entries. The module is based on the code from Peter Sagerson who gave me the [permission](https://bitbucket.org/psagers/ansible-ldap/pull-requests/2/adding-params-option-and-other-minor/diff#comment-16809629) to publish it.

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @jtyr for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by jangrewe** [comment]
shipit

Works great for me (compared to the original one, which doesn't)

**Comment by drybjed** [comment]
I have tested this module with a few Ansible roles with different options, it seems to work as expected, grat job.

shipit

File: network/ldap_entry.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+    rule allowing root to modify the server configuration. If you need to use
+    a simple bind to access your server, pass the credentials in I(bind_dn)
+    and I(bind_pw).
+version_added: '2.3'
+author:
+  - Jiri Tyr (@jtyr)
+requirements:
+  - python-ldap
+options:
+  '...':
```
Comment (bcoca) [inline network/ldap_entry.py]: we normally document this  as _raw_params

File: network/ldap_entry.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+      - If I(state=present), value or list of values to use when creating
+        the entry. It can either be a string or an actual list of
+        strings.
+  params:
+    required: false
+    default: null
+    description:
+      - Option used to allow the user to overwrite any of the other
+        options. To remove an option, set the value of the option to
+        C(null).
```
Comment (bcoca) [inline network/ldap_entry.py]: this description is very confusing

File: network/ldap_entry.py (hunk 1 above)
Comment (bcoca) [reply network/ldap_entry.py]: I would actually prefer an 'entry/object' which is a freeform dict

File: network/ldap_entry.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    objectClass: organizationalUnit
+
+- name: Make sure we have an admin user
+  ldap_entry:
+    dn: cn=admin,dc=example,dc=com
+    objectClass:
+      - simpleSecurityObject
+      - organizationalRole
+    description: An LDAP administrator
+    userPassword: "{SSHA}tabyipcHzhwESzRaGA7oQ/SDoBZQOGND"
```
Comment (bcoca) [inline network/ldap_entry.py]: see my point above, this seems confusing as the 'non existing options' are slurped into the entry definitions, there is no clear separation between the ldap entry properties and the task options. A simple misspelling will also create havok on entries.

```
prams: hello=kitty
```

**Comment by gregdek** [comment]
@jtyr A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by jtyr** [comment]
@bcoca I have fixed the things you commented on. Please feel free to review it again.

@drybjed Please could you run some more tests with the current implementation?

**Comment by drybjed** [comment]
@jtyr Nice work with the `attributes` parameter. I assume that existing roles that used the free form attribute specification will need to be fixed?

**Comment by jtyr** [comment]
@drybjed Yes, the tasks using the old version of the module must be modified. Fortunately it's quite simple - just add the `attributes` option and increase indentation of all non-module options to fall under the `attributes` option.

**Comment by drybjed** [comment]
@jtyr I have checked the new `ldap_entry` changes you added, unfortunately the module fails. The play I've used is:

```
- hosts: ldap
  become: True

  tasks:
    - name: Create main entry
      ldap_entry:
        dn: 'ou=test,dc=example,dc=com'
        objectClass: 'organizationalUnit'
```

The task fails with:

```
fatal: [ldap]: FAILED! => {"changed": false, "failed": true, "msg": "Some entry attributes must be defined."}
```

Any idea what attributes would be needed? I checked this on the `slapd` server and for organizationalUnit, top, nothing important seems to be required.

NB, this is also one of the examples in the module documentation, more or less.

**Comment by jtyr** [comment]
Thanks for running the tests, @drybjed. I have removed the requirement for having some `attributes` defined for every entry. Your example should work now.

**Comment by drybjed** [comment]
@jtyr Great, looks like it works now, after the necessary changes to tasks are applied. I'm going to switch my copy of `ldap_entry` to this version.

shipit

**Comment by jtyr** [comment]
Thanks for testing, @drybjed.

@bcoca Please could you review it now?

**Comment by gregdek** [comment]
@jtyr Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request closed without merging**